Lib/imaplib.py

```diff
@@ -1279,6 +1389,251 @@ def print_log(self):
+        self._duration = duration
+        self._imap = imap
+        self._tag = None
+        self._sock_timeout = None
+        self._old_state = None
+
+    def __enter__(self):
+        imap = self._imap
+        assert not (imap._idle_responses or imap._idle_capture)
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle start duration={self._duration}')
+
+        try:
+            # Start capturing untagged responses before sending IDLE,
+            # so we can deliver via iteration any that arrive while
+            # the IDLE command continuation request is still pending.
+            imap._idle_capture = True
+
+            self._tag = imap._command('IDLE')
+            # As with any command, the server is allowed to send us unrelated,
+            # untagged responses before acting on IDLE.  These lines will be
+            # returned by _get_response().  When the server is ready, it will
+            # send an IDLE continuation request, indicated by _get_response()
+            # returning None.  We therefore process responses in a loop until
+            # this occurs.
+            while resp := imap._get_response():
+                if imap.tagged_commands[self._tag]:
+                    raise imap.abort(f'unexpected status response: {resp}')
+
+            if __debug__ and imap.debug >= 4:
+                prompt = imap.continuation_response
+                imap._mesg(f'idle continuation prompt: {prompt}')
+        except BaseException:
+            imap._idle_capture = False
+            raise
+
+        self._sock_timeout = imap.sock.gettimeout() if imap.sock else None
+        if self._sock_timeout is not None:
+            imap.sock.settimeout(None)  # Socket timeout would break IDLE
+
+        self._old_state = imap.state
+        imap.state = 'IDLING'
+
+        return self
+
+    def __iter__(self):
+        return self
+
+    def _wait(self, timeout=None):
+        # Block until the next read operation should be attempted, either
+        # because data becomes availalable within 'timeout' seconds or
+        # because the OS cannot determine whether data is available.
+        # Return True when a blocking read() is worth trying
+        # Return False if the timeout expires while waiting
+
+        imap = self._imap
+        if timeout is None:
+            return True
+        if imap._readbuf:
+            return True
+        if timeout <= 0:
+            return False
+
+        if imap.sock:
+            fileobj = imap.sock
+        elif platform.system() == 'Windows':
+            return True  # Cannot select(); allow a possibly-blocking read
+        else:
+            fileobj = imap.readfile
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg(f'idle _wait select({timeout})')
+
+        with selectors.DefaultSelector() as sel:
+            sel.register(fileobj, selectors.EVENT_READ)
+            readables = sel.select(timeout)
```

Comment:
I haven’t really used SSLSocket.pending. It might help with the first case, if you checked it before calling _select_. It’s not clear to me how _pending_ will behave if an SSL error is pending though.

I don’t see how _pending_ would help for the second case though. After _select_ indicates the socket is readable (or writable, if the SSL library actually wants to write!), I think you’d still need to call SSL’s _recv_ (with an appropriate timeout) for it to continue its internal processing and determine if any IMAP data is ready to be read. I suspect calling _pending_ doesn’t do this processing.

Reply:
I don’t know of any other options. I would try the timeout option first.